### PR TITLE
Use Android's LruCache.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: picasso/src/main/java/com/squareup/picasso/LruCache.java [hunk 1]
```diff
@@ -19,159 +19,99 @@
...
   }
 
   @Override public void set(@NonNull String key, @NonNull Bitmap bitmap) {
     if (key == null || bitmap == null) {
       throw new NullPointerException("key == null || bitmap == null");
     }
 
-    int addedSize = Utils.getBitmapBytes(bitmap);
-    if (addedSize > maxSize) {
-      return;
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso/LruCache.java]: this was a theoretical bug . . . replacing an existing key didn’t evict if the size was too big

Comment (JakeWharton) [reply picasso/src/main/java/com/squareup/picasso/LruCache.java]: Are you sure? It was contributed because it was observed behavior.

Comment (JakeWharton) [reply picasso/src/main/java/com/squareup/picasso/LruCache.java]: Oh sorry, I'm looking at the wrong method.

Comment (JakeWharton) [reply picasso/src/main/java/com/squareup/picasso/LruCache.java]: Wait no I'm not... and you retained the behavior but added a comment?

Comment (swankjesse) [reply picasso/src/main/java/com/squareup/picasso/LruCache.java]: No, the “putting a giant image clears cache” is totally a real problem.

But there’s a theoretical problem with the fix. You have an image that’s beneath the limit in the cache. You try to update that with a higher-bytes copy of the same image. That doesn’t evict the lower-bytes predecessor, though it should.

Comment (JakeWharton) [reply picasso/src/main/java/com/squareup/picasso/LruCache.java]: Ah, cool. 👍

File: picasso/src/main/java/com/squareup/picasso/LruCache.java [hunk 2]
```diff
@@ -19,159 +19,99 @@
...
-        value = toEvict.getValue();
-        map.remove(key);
-        size -= value.allocationSize;
-        evictionCount++;
-      }
+    // If the bitmap is too big for the cache, don't even attempt to store it. Doing so will cause
+    // the cache to be cleared. Instead just evict an existing element with the same key if it
+    // exists.
+    if (byteCount > maxSize()) {
+      cache.remove(key);
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/LruCache.java]: nice, thank you. was my bug from 2 years ago. i'll update with a test.

**Comment by NightlyNexus** [comment]
I think it does want to be a public type?
People might want to configure a different size for their memory caches.
The constructor that accepts a Context seems like it shouldn't be public, though.

**Comment by swankjesse** [comment]
Probably shouldn't be a public type. I think users should configure the cache size in bytes and we should do the eviction policy.

**Comment by NightlyNexus** [comment]
Oh, I gotcha.
Don't expose Cache at all? Let users set the cache size, use the default, or set no cache when building the Picasso.

**Pull request merged**